### ISSUE TITLE
Spacemacs: Git mode causes 100 % CPU usage by treemacs-git-status.py

### ISSUE THREAD

**Issue by tsoernes** [Alexander-Miller/treemacs#379]
Opening treemacs `v2.5-2019.03.03 @ 26.1` causes 100 % CPU churn on 2 cores. I've tried setting
```
  (setq-default treemacs-git-mode nil)
  (setq-default treemacs-use-git-mode nil)
  (setq-default treemacs-use-follow-mode nil)
  (fset #'trace-follow-hook #'ignore)
```
in `dotspacemacs/user-config` and also 
```
 '(treemacs-git-mode nil)
 '(treemacs-follow-mode nil)
```
in `dotspacemacs/emacs-custom-settings` but both git-mode and follow mode gets automatically enabled
causing 100 % churn by `treemacs-git-status.py`.

**Comment by Alexander-Miller** [comment]
> (setq-default treemacs-git-mode nil)

> Setting this variable directly does not take effect;
> either customize it (see the info node `Easy Customization`)
> or call the function `treemacs-git-mode`.

> (fset #'trace-follow-hook #'ignore)

No such symbol exists, and even if it did as a hook it would've been a variable, not a function. This line does nothing.

> '(treemacs-git-mode nil)

Without context I have no idea what this does.

As for your actual issue take a look whether it's the same problem as in https://github.com/Alexander-Miller/treemacs/issues/369.

**Comment by tsoernes** [comment]
@Alexander-Miller 
```
 '(treemacs-git-mode nil)
 '(treemacs-follow-mode nil)
```
in `custom-set-variables` in `dotspacemacs/emacs-custom-settings` are the result of using `customize-mode`. **These settings are not respected**. Despite their presence (and deleting the stuff in the user-config) after rebooting Emacs, both `git-mode` and `follow-mode` are enabled (customize-mode says their value is changed outside of customize, but there is no reference to treemacs at all in the dotfile outsize of the two lines that customize has put there.

This is not the same issue as #369. After killing `treemacs-git-status.py` manually, follow-mode is still enabled while git-mode is not and treemacs works without churning CPU. Treemacs chunks out because there's 150K files in the directory. Something is very inefficient there.

**Comment by Alexander-Miller** [comment]
> in custom-set-variables in dotspacemacs/emacs-custom-settings are the result of using customize-mode. 

Those minor modes are enabled by default when treemacs is first loaded. Disabling them is done in a `:config` block with use-packge, or a plain `with-eval-after-load`. I've no idea how the customize interface gets in the way here. I suppose I can take a look, but I make no promises.

> Treemacs chunks out because there's 150K files in an unopened sub-directory of the same project. Something is very inefficient there.

Determining the files' git status is done in a parallel python process that does not block Emacs proper. The python process collects all entries and returns an elisp hash table that is fed straight into the elisp interpreter. I don't know how you have managed to accrue 150k files that have their own git status entry - other than unmodifed, which treemacs ignores - but even if you have too many lines you can still influence git-mode's performance with `treemacs-max-git-entries` and `treemacs-git-command-pipe`. So if you see something that's actually inefficient here do let me know.

**Comment by tsoernes** [comment]
The entries where not yet added to the index, so perhaps they were not
ignored by treemacs. The whole of emacs freezes when it happens; I'm unable
to move the cursor or anything. Killing the treemacs-git-status.py
processes gets Emacs responsive again; and kills treemacs-git-mode. As a
workaround I just merged the files into a single file ...

Den man. 11. mar. 2019 kl. 19:53 skrev Alexander-Miller <
notifications@github.com>:

**Comment by Alexander-Miller** [comment]
> The entries where not yet added to the index, so perhaps they were not ignored by treemacs.

I don't think there are many git systems out there that wouldn't choke on dealing with 150000 untracked files. The git cli might not, but you'd still freeze your terminal if you tried to see a detailed git status report the way treemacs does. So yeah, at that kind of scale there really is nothing I can do other than make some hard cut somewhere and just stop doing any more work, and that's exactly what those variables are for.

**Comment by tsoernes** [comment]
Git status and magit takes a second or two but works fine. In either case
it's fine if it's not able to handle that load; the problem is not being
able to disable it. Calling the function treemacs-git-mode does not give
you the option to disable it and using customize-variable does not work so
the only option is to kill the python script.

Den man. 11. mar. 2019 kl. 20:06 skrev Alexander-Miller <
notifications@github.com>:

**Comment by Alexander-Miller** [comment]
> Git status and magit takes a second or two but works fine.

Then I assume magit only sees an untracked directory, and not the single files inside. Treemacs cannot do that, it maps single files to their git status. That makes lookup in the general case very efficient, but the downside is that I need an entry for every single file. 

> the problem is not being able to disable it

Once emacs is stuck it's stuck. Best you can do is SIGUSR2. Outside of that calling `treemacs-git-mode` when it's enabled disables it, otherwise you get a dialogue for the specific git-mode type.

**Comment by vv111y** [comment]
I got this today. With 202,597 image files in a subdirectory. Emacs doesn't freeze, but my whole system is choking. I have a xeon with 8 cores, and they were all being used. I think because @tsoernes is using a 2-core laptop, that's why emacs is freezing for him - just a guess though.   
This will keep coming up in the future, unless people make a habit of keeping data dirs somewhere else. The better choice is to have some hard cut-off as you say. Did you say there is such a variable?

**Comment by Alexander-Miller** [comment]
The cut-off is `treemacs-max-git-entries`, by default it's set to 5000. Normally it should be enough to keep all your cores from being clogged with working through those 200k images, but there is a part of the script that recurses into untracked directories to get a status entry for every single untracked file, that's probably why the limit is badly applied in your case.

In addition there's `treemacs-git-command-pipe` that can be used to filter the output with something like grep.

I could also reschedule the application of a deferred git-mode if it notices that the git process is not yet finished (the default 0.5s delay should be more tha enough to cover *normal* cases), but that would just fight the symptom, not the cause. My options, however, are limited here.

I am looking at `git status` recursively because in case of a refresh of the view (which works by basically closing and reopening everything) that git status data can be passed down recursively and I need to start only a single process. Giving that up would worsen the *general* case, never mind the amount of refactoring necessary.

I can try to stop recursing into a directory that isn't/won't be visible, but I don't know how well that will work. The only other option I can see working is a libgit-based solution. There's a package for it already, but it'd be better for me to make my own thing. I could tailor the *native* parts of the code to my specific needs. I also have already experimented with a simple prototype for another part of treemacs (fetching, sorting and filtering a directory's content) that could make *that* part of the render process nearly 5 times faster. The downside, apart from the general need for compiling a binary, is that it'd have to be in Rust. It's not exactly is ubiquitous as gcc, but it is the only low-level language I am reasonably comfortable with, and cargo does make it utterly trivial to use the same optimized directory-walker libraries that ripgrep is based on.

Long story short: so much to do, so little time.

**Comment by Alexander-Miller** [comment]
See #387 & #388

**Comment by vv111y** [comment]
Thanks for your detailed response. I've subbed to both those issues

**Comment by Alexander-Miller** [comment]
It would help, yes. However it would break filewatch-mode for the linked files. Treemacs watches the linked path, but gets informed about a change in the real path, so it doesn't update anything.

**Comment by Alexander-Miller** [comment]
#387 & #388 have just been pushed. Now I need to know whether these fixes address your performance issues and whether anything else broke in the process.

**Comment by vv111y** [comment]
I am symlinking now. Do you want me to test by moving the subdir back?

**Comment by Alexander-Miller** [comment]
Please do. I tested against my own 5000 file performance test directory that is now properly ignored unless expanded, but if the fix is to be any good it must work for you as well.

**Comment by vv111y** [comment]
The initial problem is ok. But anytime (not just first time) the project is TAB-opened, emacs freezes for ~30-40sec.

**Comment by Alexander-Miller** [comment]
You will need to debug now:
 - What's the output of `git status --porcelain --ignored .` in your project root? 
 - What's the value of the `command` variable in `treemacs--git-status-process-extended` that's passed to `pfuture-new`.
 - What's the size and content of the hash table returned in `treemacs--parse-git-status-extended`?

**Comment by vv111y** [comment]
My apologies for not getting to this, just quit paranoid about looming deadlines. I will be more prompt in the future. If you still need feedback from me I can do it early May.

**Comment by Alexander-Miller** [comment]
Simple things first: verify that treemacs is actually using its smarter git features: check that `treemaca-git-mode` is set to either `extended` or `deferred` (unless you have explicitly configured it otherwise).

Then check if the issue still persists with an up-to-date treemacs. Thanks to @Jasu 's great contributions git processing recently saw even more performance pit falls averted and potential bugs fixed.

If your git-mode is set up right and it is still slow then I will need the debug info from my previous comment.
